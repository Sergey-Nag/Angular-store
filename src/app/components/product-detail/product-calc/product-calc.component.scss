.product-calc {
    line-height: 300%;
}