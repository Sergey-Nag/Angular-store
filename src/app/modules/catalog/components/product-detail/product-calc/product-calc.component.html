<div class="card product-calc p-4">
    <div class="row">
        <div class="col">Price, $</div>
        <div class="col-4 text-end">40.0</div>
    </div>
    <div class="row">
        <div class="col">Price, $</div>
        <div class="col-4">
            <input type="number" class="form-control form-control-sm">
        </div>
    </div>
    <div class="row">
        <div class="col">
            <b>Total price, $</b>
        </div>
        <div class="col-4 text-end">40.0</div>
    </div>
</div>