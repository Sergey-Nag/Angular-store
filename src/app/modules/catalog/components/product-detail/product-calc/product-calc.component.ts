import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-product-calc',
  templateUrl: './product-calc.component.html',
  styleUrls: ['./product-calc.component.scss']
})
export class ProductCalcComponent {

  constructor() { }

}
