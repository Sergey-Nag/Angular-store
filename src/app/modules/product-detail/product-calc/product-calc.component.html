<div class="card product-calc p-4">
    <div class="row">
        <div class="col">Price, $</div>
        <div class="col-4 text-end">{{price}}</div>
    </div>
    <div class="row align-items-center">
        <div class="col">Count</div>
        <div class="col-4">
            <input
                type="number"
                class="form-control form-control-sm"
                min="0"
                [max]="available"
                [value]="count"
                (change)="handleCountChange($event)">
        </div>
    </div>
    <div class="row">
        <div class="col">
            <b>Total price, $</b>
        </div>
        <div class="col-4 text-end">{{totalPrice}}</div>
    </div>
</div>