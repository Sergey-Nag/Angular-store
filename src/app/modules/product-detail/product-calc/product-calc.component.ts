import { ChangeDetectionStrategy, Component, EventEmitter, Input, OnInit, Output } from '@angular/core';

@Component({
  selector: 'app-product-calc',
  templateUrl: './product-calc.component.html',
  styleUrls: ['./product-calc.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ProductCalcComponent {
  @Input() price: number;
  @Input() available: number;
  @Input() count: number;
  @Output() countChange = new EventEmitter<number>();

  get totalPrice() {
    return (this.price * this.count).toFixed(2);
  }

  onCountChange(e: Event) {
    this.countChange.emit(this.count);
  }
}
