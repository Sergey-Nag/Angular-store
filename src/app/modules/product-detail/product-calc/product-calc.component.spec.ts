import { APP_BASE_HREF } from "@angular/common";
import { Book } from "@shared/models/book.model";
import { fireEvent, render, screen, RenderResult, } from '@testing-library/angular';
import { ProductCalcComponent } from "./product-calc.component";

describe('Product details', () => {
  let price = 56.50;
  let available = 10;
  let rendered: RenderResult<ProductCalcComponent>;
  let onCountChange: jest.SpyInstance<void, [value?: number]>

  beforeEach(async () => {
    rendered = await render(ProductCalcComponent, {
      componentProperties: {
        price,
        available,
      }
    });

    onCountChange = jest.spyOn(rendered.fixture.componentInstance.onCountChange, 'emit');
  });

  it('Should render a table', () => {
    expect(rendered.fixture).toMatchSnapshot();
  });

  it('Should calculate total price and emit the onCountChange event when count value changes', () => {
    fireEvent.change(screen.getByRole('spinbutton'), { target: { value: '2' }});

    expect(screen.getByText(`${(price * 2).toFixed(2)}`)).toBeTruthy();
    expect(onCountChange).toBeCalledWith(2);
  });

  it('Should emit the onCountChange event with 0 if the input value less than 0', () => {
    fireEvent.change(screen.getByRole('spinbutton'), { target: { value: '-5' }});
    
    expect(onCountChange).toBeCalledWith(0);
  });

  it('Should emit the onCountChange event with max available value if the count value more than available', () => {
    fireEvent.change(screen.getByRole('spinbutton'), { target: { value: '25' }});

    expect(onCountChange).toBeCalledWith(available);
  });
});