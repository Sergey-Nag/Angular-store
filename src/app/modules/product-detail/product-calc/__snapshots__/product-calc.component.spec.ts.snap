// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product details Should render a table 1`] = `
<app-product-calc
  __ngContext__={[Function LRootView]}
  count={[Function Number]}
  onCountChange={[Function EventEmitter_]}
>
  <div
    class="card product-calc p-4"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        Price, $
      </div>
      <div
        class="col-4 text-end"
      >
        56.5
      </div>
    </div>
    <div
      class="row align-items-center"
    >
      <div
        class="col"
      >
        Count
      </div>
      <div
        class="col-4"
      >
        <input
          class="form-control form-control-sm"
          max="10"
          min="0"
          type="number"
        />
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <b>
          Total price, $
        </b>
      </div>
      <div
        class="col-4 text-end"
      >
        56.50
      </div>
    </div>
  </div>
</app-product-calc>
`;
